### ISSUE TITLE
Issue with nifty-gui text rendering with LWJGL

### ISSUE THREAD

**Issue by Tutzen** [nifty-gui/nifty-gui#312]
I'm feeling bad for posting the third issue within a few days that's not actually an issue but more me not really understanding how everything works, but I don't know where else to get help. I posted this in my last issue, but since it's already closed you probably didn't see it.

OpenGL works fine now, like I said - but now I've got problems with the rendering of nifty in my actual game window. It looks like this:

![nifty-text](https://cloud.githubusercontent.com/assets/10613635/5966748/4de2ca10-a807-11e4-9c1f-cfa36aa61e4e.png)

Does that sound (or, to be exact, look) familiar to you? I already tried to mess around a bit with my rendering loop but if I render nifty before OpenGL, it doesn't appear on the screen at all and when I call render(true) instead of render(false), I get a completely black screen.
Can you help me once more with this issue? Didn't find an appropriate answer via Google or another issue right here yet...

Some code excerpts that may be helpful:

The screen:

``` XML
<?xml version="1.0" encoding="UTF-8"?>
<nifty xmlns="http://nifty-gui.lessvoid.com/nifty-gui">
    <useStyles filename="nifty-default-styles.xml" />
    <useControls filename="nifty-default-controls.xml" />

    <screen id="start" 
            controller="htw.gt3.trappers.client.screens.ClientGameController">
        <layer childLayout="vertical">
            <panel id="playerPointsPanel" height="100%" width="100%" 
                   align="left" childLayout="horizontal">
                <text font="aurulent-sans-16.fnt" color="#000000" 
                      text="Collisions: " textHAlign="left" />
                <text id="playerPoints" font="aurulent-sans-16.fnt" 
                      color="#000000" text="0" textHAlign="left" />
            </panel>
        </layer>
    </screen>

</nifty>
```

The rendering loop:

``` Java
while(!Display.isCloseRequested()) {
      Controls.checkInput();
      processMessages();
      if(prepares) {
            createSegment();
            createTraps();
            player.setSpeed(player.RUNNINGSPEED);
            prepares = false;
      }
      if(started) {
            Graphics.render();
            UI.getNifty().update();
            glPushMatrix();
            glPushAttrib(GL_ALL_ATTRIB_BITS);
            glEnable(GL_BLEND);
            UI.getNifty().render(false);
            glDisable(GL_BLEND);
            glPopAttrib();
            glPopMatrix();
            glDisable(GL_TEXTURE_2D);
       } else {
             UI.getNifty().update();
             glEnable(GL_BLEND);
             UI.getNifty().render(true);
             glDisable(GL_BLEND);
       }
       Display.update();
       Display.sync(60);
}
```

I hope it doesn't have anything to do with the new render factory I'm using? It works fine in screens without actual OpenGL objects where I don't render OpenGL at all, the issue only starts when there is OpenGL drawn onto the screen.

**Comment by spaceemotion** [comment]
For me it seems like you're not setting your blend modes at all. It currently looks like its using either a wrong, or no one. For a nice chart of the blending modes you can look at this image: http://www.quake-1.com/docs/blending.jpg.

**Comment by void256** [comment]
Yes, what @spaceemotion said and you want to check the example code for the states required for Nifty rendering as well. Here is a link: https://github.com/void256/nifty-gui/blob/1.4/nifty-examples-lwjgl/src/main/java/de/lessvoid/nifty/examples/lwjgl/NiftyExampleLoaderLWJGL.java#L245  <-- you want to have to set these states set before you call nifty.render().

Edit: I've [updated the wiki](https://github.com/void256/nifty-gui/wiki/How-to-save-OpenGL-state-when-render-with-Nifty) as well.

**Comment by Tutzen** [comment]
Okay, I inserted the complete code example from the Wiki in my loop, right after the update and right before the render call. The problem remains, though, unfortunately. Do you have any other suggestions?

Setting the blend mode didn't solve it, too.

**Comment by void256** [comment]
Ok, I've took a closer look and it seems I'm not up-to-date with Nifty :wink: Some time ago there was a big refactoring done by Aaron to integrate batched rendering even for the old way of using OpenGL. What I was not aware - until now - is, that we have the following code inside the batched renderer:

``` java
  private void beginRendering() {
    gl.glEnable(gl.GL_TEXTURE_2D());
    gl.glEnable(gl.GL_BLEND());
    gl.glEnableClientState(gl.GL_VERTEX_ARRAY());
    gl.glEnableClientState(gl.GL_COLOR_ARRAY());
    gl.glEnableClientState(gl.GL_TEXTURE_COORD_ARRAY());
  }

  private void endRendering() {
    gl.glDisableClientState(gl.GL_TEXTURE_COORD_ARRAY());
    gl.glDisableClientState(gl.GL_COLOR_ARRAY());
    gl.glDisableClientState(gl.GL_VERTEX_ARRAY());
    gl.glDisable(gl.GL_BLEND());
    gl.glDisable(gl.GL_TEXTURE_2D());
  }
```

So, actually enabling blending or texturing is not necessary anymore, since Nifty will do that automatically in each render() call.

Do you use any of the clienState in your application?

It might be necessary to add `glPushClientAttrib(GL_CLIENT_ALL_ATTRIB_BITS)` before rendering and `glPopClientAttrib()` after rendering.

If I remember correctly you get Nifty rendering working when you are not rendering stuff on your own and only Nifty is rendering. Is this still correct? If so, then there has to be some OpenGL state modified that you're not saving.

If this is not correct then we should get Nifty first working stand-alone.

Do you use any multi texturing?

A SSCCE would help in all the case ... :wink:

**Comment by Tutzen** [comment]
Got it working. Sorry, guys, it was just a very stupid fault.

I have a 400x200 display at the beginning because there the client has to connect to the server and stuff. I then change the resolution of the screen to 800x600, but I did that by first destroying it and then creating a new one. Obviously nifty got problems with that since I didn't re-initialise the UI afterwards, I only informed nifty about the resolution change. I don't delete the screen anymore, I just set the displayMode again and it works now.

Stupid me, I guess. Thanks for your help anyway, you're amazing. :)
